### ISSUE TITLE
Feature request: Soundtouch multiroom

### ISSUE THREAD

**Issue by mkoninkx** [kalkih/mini-media-player#155]
Hi!

I'm a big fan of your component! Thanks for all the time invested in it!
Would it be possible to integrate multiroom support for Bose Soundtouch speakers?
Behaviour would be the same as with Sonos.

Many thanks!
Br
Martijn

**Comment by kalkih** [comment]
Hello, Thanks.

Yes, from what I can tell after looking at the documentation it should be possible.

The service calls are not exactly the same as Sonos so would require some conditional statements, but def possible to implement.

**Comment by mkoninkx** [comment]
Hi Karl,

Great, thanks!
Let me know if you need a beta tester, I've got 3 soundtouch speakers setup in Home Assistant.

**Comment by da-anda** [comment]
I'd also be interested in it. Since you already seem to do some special casing for `sonos` related join/unjoin calls, would you be fine with doing the same for `soundtouch`? Alternatively it might be possible to make the core code more generic and build the service call structure out of some platform definitions which also could be overridden in the player config. Not sure how complicated such a "templating" kinda thing would be inside a lovelace component

**Comment by kalkih** [comment]
Adding an exception for the platform would probably be the way to go, similar to how we did with sonos before.
The sonos way now seems to be the "standard" way going forward with the format of `platform/action`.
See recent changes #161

Do you know if [`SOUNDTOUCH_ADD_ZONE_SLAVE`](https://www.home-assistant.io/components/soundtouch/#service-soundtouch_add_zone_slave) will create the group if it doesn't exist, similar to how [other platforms](https://www.home-assistant.io/components/sonos/#service-sonosjoin) does it?

**Comment by kalkih** [comment]
Hey, I've opened a PR with some initial code for the implementation #165.
Would be great if you guys could try it out. I've compiled a version of the PR branch that can be found [here (gist)](https://gist.github.com/kalkih/7f6161ca2e21cdbc66ac85cb6b9264e5).

Example config would look something like this:
```yaml
- type: custom:mini-media-player
  entity: media_player.office
  speaker_group:
    platform: soundtouch
    show_group_count: true
    entities:
      - entity_id: media_player.office
        name: Office
      - entity_id: media_player.livingroom
        name: Livingroom
      - entity_id: media_player.kitchen
        name: Kitchen
```

Thanks!

**Comment by da-anda** [comment]
> Do you know if SOUNDTOUCH_ADD_ZONE_SLAVE will create the group if it doesn't exist, similar to how other platforms does it?

Just tested it by manually calling the service and it returns an exception that the zone doesn't exist. So unfortunately not :(

**Comment by kalkih** [comment]
Okay, shouldn't really be an issue, although, if the component doesn't update its state fast enough it could cause some delay/issues after grouping the first entity (creating the group) and adding additional entities quickly after.

https://github.com/kalkih/mini-media-player/blob/db743f2969a28b6c79f0e1baec8aab4598d6494e/src/model.js#L297

**Comment by da-anda** [comment]
I think looping the selected speakers might become an issue, but rewriting the player logic to make use of the possibility to send multiple slave IDs to the Soundtouch service is probably also not an option. Too bad that there is no standard across the components (at least not yet).

**Comment by mkoninkx** [comment]
Thanks for the work done so far! Please contact me when I need to test something.

**Comment by jpearce73** [comment]
Hi There.
I've been working through Bose Soundtouch HA for a little while now, looking to improve that whole experience. Your card is going to be a big part of that!

So far I have a lot of local work in custom_componentns which does
1. Upgraded libsoundtouch 0.7.2 => 0.8.0, 
2. Made some necessary folow on changes to media_player/soundtouch
3. Made further modification to libsoundtouch as Bose API seems to have been changing and libsoundtouch does not seem to have any active maintenance (git page has gone dead). One of these modifications implements ADD_ZONE_SLAVE with SET_ZONE, which works the way you want in later version of Bose API. It creates the zone if requires and adds to it if its already created.

=> All this is still under initial testing, but I am finally able to call soundtouch services to create and add/remove devices from zones.

Now I want to connect to this card and I am trying to compile up your component. I followed the developer instructions but am stuck on the Node.js part. I gather I need to install Node.js on HASS.io to get the npn install to work. Any pointers on how to do this?

**Comment by kalkih** [comment]
Hey, that's cool!
The main thing that keeps the speaker group management card feature from not working with the official soundtouch component is the lack of an exposed `group` attribute, we need this to keep track of groups and group members/master etc.

In order to function without any modifications to this card the `group` attribute should look something like this (this is also how the Sonos component does it):

- The group is a list/array.
- The group contains the entity id's of all of its members.
- The group looks the same on master and client/slave.
- The initial entry is the master.
- The attribute is exposed as`[platform]_group` (`soundtouch_group` in this case).

Yes, npm is bundled with node.js.
You don't necessarily have to install node.js on your hass.io machine, I'm personally not running hass.io so I don't know how you would go about it, but you could always compile the code on another machine and then move over the file(s).

**Comment by jpearce73** [comment]
Thanks.
Nice instructions in the developer section of your docs too. I know nothing about node.js and I was able to build the dist target by following your steps (details below).
Now I'll see if I can get this going...

For anyone looking to do this on a Win10 box.
1. I installed Ubuntu Linux from the microsoft store.
2. Follow the instructions about enabling the Linux subsystem
3. apt-get seems to be broken out of the box, so
  a) >sudo apt-get update
  b) >sudo apt-get upgrade
4. Install node.js
   a) > curl -sL https://deb.nodesource.com/setup_12.x | sudo -E bash -
   b) > sudo apt-get install nodejs
   c) > node -v
      v12.10.0
   d) > npm -v
      6.10.3
5 Follow component docs to build  dist/mini-media-player-bundle.js
6. copy to HA (assumes HASS.io)
  a) scp dist/mini-media-player-bundle.js root@<HASSIO_IP>:/config/www/

**Comment by da-anda** [comment]
you don't need the linux subsystem just to run node.js. There is also a Windows installer for it. I personally use `chocolatey`(https://chocolatey.org/) which is a Windows package manager similar to `aptitude` on debian. Works fine for most stuff.

**Comment by jpearce73** [comment]
I'm making some progress - but probably need some help...

I added a property to soundtouch/media_player.

```
   @property
    def group(self):
        """Return the group status of the device in format compatible with mini-media-player"""
        _LOGGER.debug("Returning property group: %s", self._zone_status)
        ret = []
        ret.append(self._zone_status._master_id)
        ret.append(self._zone_status._slaves)
        return ret
```
I was expecting to be able to see this new attribute in HA's Developer=>States page. But it doesnt show up. What controls what is seen there?

Is there some logging mechanism for the java script?
You call soundtouch_group in model.js - can I dump a log message so I can confirm what happening?

**Comment by kalkih** [comment]
Hey, I would recommend having a look at how other components does it, check the [Sonos component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sonos/media_player.py#L1198-L1209) for example.

Yes, you can log things in javascript with the help of [console.log()](https://developer.mozilla.org/en-US/docs/Web/API/Console/log, check the browser developer console for the output.

**Comment by jpearce73** [comment]
I figured out that I need to add a device_state_attributes function.
I can now see my new group attribute.

Would be good to know if I can add debug messages into the java script...

**Comment by jpearce73** [comment]
OK I have the group working ~almost~ as you described 
Let's say we have 3 devices playing in a group with the kitchen being the master.

media_player.bose_kitchen.group = [media_player.bose_kitchen, media_player.bose_lounge,media_player.bose_stairs]
media_player.bose_lounge.group = [media_player.bose_kitchen, media_player.bose_lounge]
media_player.bose_stairs.group = [media_player.bose_kitchen, media_player.bose_stairs]

=> The issue I have is that I cannot get the slaves to know about the other slaves in the group. Each slave knows its master but does not know about the other slaves. This is because the bose API does not return that information. And once I bring the information into HA, is all split up in the different entities. 

When I use the mini-media-player 
- I can click the checkbox and create a zone and it works
- I can click additional checkboxes and add to a zone and it works
- unchecking a box calls add_zone_slave instead or remove_zone_slave, which does nothing
- the media card does not track changes made elsewhere, i.e. If I use the bose app to add/remove from the zone, it is not seen by the media card (it is seen soundtouch_group attributes in Developer Tools => States)

I need to look into the java script now...

**Comment by jpearce73** [comment]
I am a bit stuck here. :(
As described above, I have exposed SOUNDTOUCH_GROUP, plus a bonus attribute IS_MASTER.
![image](https://user-images.githubusercontent.com/19938737/66272582-630e7480-e820-11e9-86c9-f2295737de41.png)



I do not understand why the min-card code below will not call SOUNDTOUCH_REMOVE_ZONE_SLAVE.
I can create a zone OK and add more slaves and I have checked boxes in the mini card.
But when I uncheck a speaker it calls SOUNDTOUCH_ADD_ZONE_SLAVE again!
I even went back to the original code before pull request #165 and implemented SOUNDTOUCH_JOIN / SOUNDTOUCH_UNJOIN in the soundtouch library. This gives it a API compatible with SONOS for example - but the mini-card always calls JOIN, never UNJOIN.
I would be expecting "checked" to be passed in as the state of the check box and it would be a very straight forward brank at if (checked).

```
handleGroupChange(e, entity, checked) {
    const { platform } = this.config.speaker_group;
    const options = { entity_id: entity };
    if (checked) {
      console.log('handleGroupChange - checked');
      options.master = this.config.entity;
      if (platform === 'bluesound') {
        return this.callService(e, `${platform}_JOIN`, options);
      }
      if (platform === 'soundtouch') {
        const service = this.isGrouped ? 'SOUNDTOUCH_ADD_ZONE_SLAVE' : 'SOUNDTOUCH_CREATE_ZONE';
        return this.handleSoundtouch(e, service, entity);
      }
      this.callService(e, 'join', options, platform);
    } else {
      console.log('handleGroupChange - unchecked');
      if (platform === 'bluesound') {
        return this.callService(e, `${platform}_UNJOIN`, options);
      }
      if (platform === 'soundtouch') {
        return this.handleSoundtouch(e, 'SOUNDTOUCH_REMOVE_ZONE_SLAVE', entity);
      }
      this.callService(e, 'unjoin', options, platform);
    }
  }

I couldn't get that log messaging to work either. When I compile, I get these warnings and I do not get any messages in the Chrome console:
  292:7  warning  Unexpected console statement  no-console
  303:7  warning  Unexpected console statement  no-console

```

Here's a look at my HA logging. I am checking and then unchecking the master_bedroom speakers within a  few seconds and it calls SOUNDTOUCH_JOIN both times.

```
2019-10-06 09:51:14 DEBUG (SyncWorker_16) [custom_components.soundtouch.media_player] service_handle <ServiceCall media_player.soundtouch_join (c:691c029685c542c09004be0afe37c94d): entity_id=['media_player.bose_master_bedroom'], master=media_player.bose_kitchen> with master=media_player.bose_kitchen, slave_ids=media_player.bose_master_bedroom
2019-10-06 09:51:14 INFO (SyncWorker_16) [custom_components.soundtouch.media_player] Adding slaves media_player.bose_master_bedroom to zone with master media_player.bose_kitchen
2019-10-06 09:51:14 DEBUG (SyncWorker_16) [libsoundtouch.device] request.post url=http://192.168.2.180:8090/addZoneSlave, payload=<?xml version="1.0" encoding="UTF-8" ?><zone master="000C8AC06D11"><member ipaddress="192.168.2.185">F45EAB5367D7</member></zone>, response=<?xml version="1.0" encoding="UTF-8" ?><status>/addZoneSlave</status>
2019-10-06 09:51:16 DEBUG (SyncWorker_7) [custom_components.soundtouch.media_player] service_handle <ServiceCall media_player.soundtouch_join (c:18aafe7c2da240948050484a52713aca): entity_id=['media_player.bose_master_bedroom'], master=media_player.bose_kitchen> with master=media_player.bose_kitchen, slave_ids=media_player.bose_master_bedroom
2019-10-06 09:51:16 INFO (SyncWorker_7) [custom_components.soundtouch.media_player] Adding slaves media_player.bose_master_bedroom to zone with master media_player.bose_kitchen
2019-10-06 09:51:16 DEBUG (SyncWorker_7) [libsoundtouch.device] request.post url=http://192.168.2.180:8090/addZoneSlave, payload=<?xml version="1.0" encoding="UTF-8" ?><zone master="000C8AC06D11"><member ipaddress="192.168.2.185">F45EAB5367D7</member></zone>, response=<?xml version="1.0" encoding="UTF-8" ?><status>/addZoneSlave</status>

```

**Comment by kalkih** [comment]
@jpearce73 Glad you're still working on this!

Okay, so the checked property is set here:
https://github.com/kalkih/mini-media-player/blob/a3f56edfcdfb4cd6a0070fa1e43d90e76402dfeb/src/components/groupList.js#L40

`checked` will equal `true` if either of the following statements are `true`
* the entity is the card entity (`item.entity_id === this.player.id`)
* the entity exists in the group attribute (`group.includes(item.entity_id)`)

The group getter (which the latter statement is based of) can be found here:
https://github.com/kalkih/mini-media-player/blob/a3f56edfcdfb4cd6a0070fa1e43d90e76402dfeb/src/model.js#L84-L87

***Important***: The checkbox might not always be in sync with `checked` property, the checkbox is not completely synced with the backend, since that would/could cause delays of the visual updates and possibly confuse the user.
The `checked` property is however always in sync with the backend component.

The `handleGroupChange` is initiated here:
https://github.com/kalkih/mini-media-player/blob/a3f56edfcdfb4cd6a0070fa1e43d90e76402dfeb/src/components/groupItem.js#L26-L34

And will return a boolean opposite of the `checked` property (described earlier), so if the `checked` property equals `true` the `checked` parameter in the `handleGroupChange` call will be `false` and vice versa.

So if the `group` isn't picked up properly by the card the `handleGroupChange` will always be called with the parameter `checked` as `true`, and will as a result always call the join service, since the card doesn't know the entity is part of the group. **This is the most likely the cause of the issue**.

I recommend making sure that the group attribute is picked up correctly by the card:
https://github.com/kalkih/mini-media-player/blob/a3f56edfcdfb4cd6a0070fa1e43d90e76402dfeb/src/model.js#L84-L87

Console log it in the group getter for example and make sure it's an array and the entity_id's are correct etc.

Are you sure your component implementation of the group attribute expose the attribute as an array/list and not as a comma separated string?

**Comment by jpearce73** [comment]
Thanks for your detailed reply; it really helped me get to the bottom of this.
Problem: The group was not the correct data structure!
Not quite as you had guessed, but close. I had this
group = [master, [slave1,slave2,slave3]]

i.e. I was joining master and the slave list incorrectly in python and it was a list containing 2 elements, a master and then a list of slaves. But HA's state page displayed this the same way as a single list.

Now everything seems to be working well. Really well actually. :)
The state of the checkboxes even updates when I add/remove slaves via other devices.

I have no changes for mini-media-player, my only changes are debugging console statements.
I have a lot of changes to the soundtouch component and I am not really sure how to distribute this for testers. I will figure something out and post it on HA forums and link back and forth to here.

**Comment by kalkih** [comment]
Awesome!

Yes, posting it in the HA forum sounds like a good idea, are you planning to submit a PR to the core HA repo with the component changes?

**Comment by jpearce73** [comment]
Yes, I'll submit a pull request.
But there are a few barriers I have to overcome...

1. This project began a while ago, based on HA 0.96.0 release. I copied the component into custom_components and started working away. Since then soundtouch has been run through a linting / formatting tool and its totally changed the file. So I'll need to manually re-do each change in at the head.

2. Half my changes are in soundtouch within HA repo. The other half are in libsoundtouch (https://github.com/CharlesBlonde/libsoundtouch) which is a dependancy of soundtouch. The owner of that repo does not seem to be active anymore - so I am not sure how to approach that!

3. I have a lot of debugging messages in the code, which helped me develop and monitor some issues in the original libsoundtouch. I am not sure if leaving these in the pull request is permissible or not. I've done it with the _LOGGER.debug() calls but I do not see that used in other modules. But I want to keep them in my copy so I can review logs and catch anything that might go wrong in the coming weeks. I am not sure if I need to do that on a separate branch or something. 

4. On top of all of that, I am new to git and python and HA! I know older version control systems like CVS and perforce, older languages. So I have to read up on the conventions and expectations of submitting a pull request, then even work out how to do one! :)

It may take me a week or two to work all this out.

If someone is reading this, and has the know-how and wants my (raw) changes I will upload them here.
The soundtouch stuff can be placed in custom_components/soundtouch/media_player.py
But the libsoundtouch is very hard to replace, it doesn't work in /config/custom_components and needs to be in the root install of python. I actually created a symbolic link from the root install of python to custom_components/libsoundtouch, but that process is dependant on how you install HA (Hass.IO vs other methods).

[soundtouch_beta.zip](https://github.com/kalkih/mini-media-player/files/3695422/soundtouch_beta.zip)

**Comment by kalkih** [comment]
2. Hmm, not sure how you'd go about the inactive upstream repo, that's an issue.

3. Here's the HA docs repo [home-assistant.io](https://github.com/home-assistant/home-assistant.io/blob/cbe3fd28e662eb8126280f3ec2d6cbaa5e6b4d25/source/_integrations/soundtouch.markdown).

Don't be afraid of submitting PRs, they don't have to be complete or even halfway done, it's a good way to get a discussion going and get help/input from others, and you can continue from there.

**Comment by da-anda** [comment]
FYI - I updated my HA PR referenced above. This should fix multiroom handling. What I noticed though is, that there seems to be some kind of API bug in libsoundtouch (or the BOSE API) because I was not able to destroy a group in one go. So the "ungroup" button which is sending all slaves to the `remove_zone_slave` method basically did nothing in my case, even though the event was triggered.

I also noticed that when you join a speaker to an existing group, for some reason the master is set wrong in the API (API response of speaker is showing itself as master). Haven't found out yet why this happens - but doesn't look like it's a bug on our end.

**Comment by jpearce73** [comment]
Hey there.
After moving house (and countries) I've got myself a new HA instance and I installed my sound touch system with a clean HA install. I found that the slave lookup had a bug. I've been trying to do a proper development this time but am having real trouble getting an environment going.
Any help from this group would be appreciated!

I've got this thread
https://community.home-assistant.io/t/looking-for-a-simple-method-to-debug-and-enhance-a-component/184696

and this pull request for my soundtouch fix that I cannot reconcile with the submission rules
https://github.com/home-assistant/core/pull/34037
